lambdas/cli.py

```diff
@@ -189,48 +200,60 @@ def bulk_validate(
 
     # initialize results dataframe with input data
     results_df = input_df.copy()
+    results_df["aip_s3_uri"] = None
     results_df["valid"] = False
     results_df["error"] = None
+    results_df["error_details"] = None
     results_df["elapsed"] = None
 
-    results_lock = Lock()
+    # handle pre-existing results
+    skip_aip_uuids = []
+    if os.path.exists(output_csv_filepath):
 
-    def validate_worker(index: Any, row: pd.Series) -> None:  # noqa: ANN401
-        """Worker function to validate a single AIP and update results DataFrame."""
-        aip_uuid = row.get("aip_uuid")
-        s3_uri = row.get("aip_s3_uri")
-
-        if not (aip_uuid or s3_uri):
-            error_msg = "Row must have either aip_uuid or aip_s3_uri"
-            logger.error(error_msg)
-            with results_lock:
-                results_df.loc[index, "error"] = error_msg
-            return
-
-        try:
-            result = validate_aip_via_lambda(
-                aip_uuid=aip_uuid, aip_s3_uri=s3_uri, verbose=ctx.obj["VERBOSE"]
-            )
+        # update results with pre-existing results
+        existing_results_df = pd.read_csv(output_csv_filepath)
+        existing_results_indexed = existing_results_df.set_index("aip_uuid")
+        results_df_indexed = results_df.set_index("aip_uuid")
+        results_df_indexed.update(existing_results_indexed)
+        results_df = results_df_indexed.reset_index()
+
+        # prepare list of AIPs to skip
+        if retry_failed:
+            skip_aip_uuids = list(results_df[results_df.valid].aip_uuid)
+        else:
+            skip_aip_uuids = list(results_df[~results_df.valid.isna()].aip_uuid)
+        logger.info(
+            f"Found {len(skip_aip_uuids)} already validated AIPs, will skip these."
+        )
 
-            # update results dataframe for AIP
-            with results_lock:
-                results_df.loc[index, "valid"] = bool(result.get("valid", False))
-                results_df.loc[index, "error"] = result.get("error")
-                results_df.loc[index, "elapsed"] = result.get("elapsed")
-                results_df.loc[index, "aip_s3_uri"] = result.get("s3_uri")
+    # init the output CSV file with previous results (if any)
+    results_df[
+        (results_df.aip_uuid.isin(skip_aip_uuids))
+        & (results_df.aip_uuid.isin(input_df.aip_uuid))
+    ].to_csv(output_csv_filepath, index=False)
```

Comment:
Perhaps related to my previous comment re: understanding what the _output_ CSV file looks like. 🤔 Is there any risk to overwriting the contents of the output CSV file _before_ the run is complete? Does the output CSV file only include rows for AIPs that were validated?

Reply:
It's related to your other question I think, insofar as it relates to figuring out that skip list.  The skip list of AIPs is kind of central to how this all works.  And, that a user is very conciously providing the **same output CSV filepath**.

> Does the output CSV file only include rows for AIPs that were validated?

After this block of code is run, that's correct.  Any AIPs that were present in the input CSV and part of the skip list, get copied over exactly as they were from the previous run.  

> Is there any risk to overwriting the contents of the output CSV file before the run is complete? 

In a sense, this happens everytime!  The relevant bits of the previous results are copied over, but it overwrites the file at that same moment.